src/main/java/org/truffleruby/core/FinalizationService.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2013, 2017 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.core;
+
+import com.oracle.truffle.api.object.DynamicObject;
+import org.truffleruby.RubyContext;
+import org.truffleruby.core.thread.ThreadManager;
+
+import java.lang.ref.ReferenceQueue;
+import java.lang.ref.WeakReference;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class FinalizationService {
+
+    private static class Finalizer {
+
+        private final Class<?> owner;
+        private final Runnable action;
+        private final Supplier<Stream<DynamicObject>> roots;
+
+        public Finalizer(Class<?> owner, Runnable action, Supplier<Stream<DynamicObject>> roots) {
+            this.owner = owner;
+            this.action = action;
+            this.roots = roots;
+        }
+
+        public Class<?> getOwner() {
+            return owner;
+        }
+
+        public Runnable getAction() {
+            return action;
+        }
+
+        public Supplier<Stream<DynamicObject>> getRoots() {
+            return roots;
+        }
+    }
+
+    private static class FinalizerReference extends WeakReference<Object> {
+
+        public List<Finalizer> finalizers = new LinkedList<>();
+
+        public FinalizerReference(Object object, ReferenceQueue<? super Object> queue) {
+            super(object, queue);
+        }
+
+        public void addFinalizer(Class<?> owner, Runnable action, Supplier<Stream<DynamicObject>> roots) {
+            finalizers.add(new Finalizer(owner, action, roots));
+        }
+
+        public void removeFinalizers(Class<?> owner) {
+            finalizers.removeIf((f) -> f.getOwner() == owner);
```

Comment:
FWIW lambdas with a single argument can be written without the extra ():
```java
finalizers.removeIf(f -> f.getOwner() == owner);
```
I find it more readable.